fix the bug about seq (issue #947)

LvHang opened the pull request:
@danpovey @vince62s 
This pull request is about issue #947 
1.I think vince62s's problem is mainly caused by the difference between 'dash' and 'bash' when they try to  interpret `{}` , rather than ‘seq’.
For example, the command "lat.{1,2,3}.gz" will be replaced by "lat.1.gz lat.2.gz lat.3.gz" in bash, but nothing will be done in dash.
2.I submit a solution and I have tested them in both dash and bash. Meanwhile, I did some tests on Switchboard. I think everything goes smoothly.
  ```C++
 - lats=$(eval echo $dir/lat.{$(seq -s, $nj)}.gz)
 + for n in $(seq $nj); do
 +   lats=$(eval echo "$lats $dir/lat.$n.gz")
 + done
  ```
3.I still have a doubt. At the begining of "get_ctm.sh" file, there is a command "#!/bin/bash ". I think the script will be conducted in bash circumstance. Why vince62s's problem will happen? I'm not very clear about it. 

If the code has any questions, please contect me as soon as possible. I would appreciate it very much.
Hang

vince62s commented:
I fully agree, I am pretty sure it was run in bash .....
and the error message was that lat.{1}.gz did not exist, as if it was unable to handle the { }
but there are many other instances where the array stuff { } is properly interpreted.

jtrmal commented:
It was discussed here or on the mailing list a couple of times. The issue
 that popen(3) and all the higher level functions based on this call use sh
to run the command. Depending on to what sh is linked to --either dash or
bash -- the {} expansion is done correctly or not.

quoting popen(3):
The *command* argument is a pointer to a null-terminated string containing
a shell command line. This command is passed to*/bin/sh* using the *-c* flag;
interpretation, if any, is performed by the shell.


y.

On Fri, Aug 12, 2016 at 4:32 PM, vince62s <notifications@github.com> wrote:

> I fully agree, I am pretty sure it was run in bash .....
> and the error message was that lat.{1}.gz did not exist, as if it was
> unable to handle the { }
> but there are many other instances where the array stuff { } is properly
> interpreted.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/963#issuecomment-239462323>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKisX8QbA3kOXWU2x0ZNLblCQeWA6L_aks5qfIPwgaJpZM4JjFTY>
> .
>

jtrmal commented:
So the differences in behavior you've seen basically boils down to the
question where the expansion is supposed to happen -- if in the script
directly, then it will work ok (as the bash was specified in the hashbang),
but if you submit the expansion as a part of the ark: specification, it
will depend on what sh is symlinked to.
y.

On Fri, Aug 12, 2016 at 4:45 PM, Jan Trmal <jtrmal@gmail.com> wrote:

> It was discussed here or on the mailing list a couple of times. The issue
>  that popen(3) and all the higher level functions based on this call use sh
> to run the command. Depending on to what sh is linked to --either dash or
> bash -- the {} expansion is done correctly or not.
>
> quoting popen(3):
> The *command* argument is a pointer to a null-terminated string
> containing a shell command line. This command is passed to*/bin/sh* using
> the *-c* flag; interpretation, if any, is performed by the shell.
>
>
> y.
>
> On Fri, Aug 12, 2016 at 4:32 PM, vince62s <notifications@github.com>
> wrote:
>
>> I fully agree, I am pretty sure it was run in bash .....
>> and the error message was that lat.{1}.gz did not exist, as if it was
>> unable to handle the { }
>> but there are many other instances where the array stuff { } is properly
>> interpreted.
>>
>> —
>> You are receiving this because you are subscribed to this thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/kaldi-asr/kaldi/pull/963#issuecomment-239462323>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/AKisX8QbA3kOXWU2x0ZNLblCQeWA6L_aks5qfIPwgaJpZM4JjFTY>
>> .
>>
>
>

vince62s commented:
ok very clear, thanks.

danpovey commented:
There are two separate issues: the bash/dash issue, and the fact that it won't work if nj == 1.   They can both be fixed in the same way, but please note that we're not talking about fixing this in just one place, there are many places in the scripts where this needs to be fixed, you need to do a 'find'.
Also, you solution takes O(nj^2), which is too slow.
You should do
```
 lats=$(for n in $(seq $nj); do echo $dir/lat.$nj.gz; done)
```
And if any of these scripts check whether /bin/sh links to dash, remove that check.

Dan

LvHang commented:
@jtrmal 
Your explanation is very clear. Thank you very much.

@danpovey 
I searched all the 'seq -s' commands in kaldi. I fixed most of them with the similar "for loop" commands as your suggestion. And if any of these scripts check whether /bin/sh links to dash, I have removed it.
The remaining two are
1. egs/csj/s5/local/csj_make_trans/csj_autorun.sh:
```C++
"dvd" ) num=dvd  ; SDB=  ; WAV= ; disc=$num`seq -s " "$num 3 17` ;;
```
Because I think it is safe because it will work in the script.
2. egs/librispeech/s5/local/run_rnnlm.sh:
```C++
rnnlm_cmd="taskset -c $(seq -s, 0 $(( $num_threads - 1 )) ) $rnnlm_cmd"
```
Because I checked the usage and found that *taskset* needs a comma-separated list. Meanwhile, I think it is safe,too.

Thank you very much for your guidance.
Hang

danpovey commented on egs/librispeech/s5/local/lm/train_lm.sh:
```diff
@@ -47,7 +47,7 @@ split_prefix=$tmp_dir/split
   mkdir -p $tmp_dir
   echo "Splitting into $normjobs parts, to allow for parallel processing ..."
-  split_files=$(eval "echo $split_prefix-{$(seq -s',' $normjobs)}")
+  split_files=$(for n in $(seq $normjobs); do echo $split_prefix-$n; done)
```
I think it's better if you revert this change, that will avoid me having to verify that it's correct.

danpovey commented on egs/librispeech/s5/local/prepare_dict.sh:
```diff
@@ -72,7 +72,7 @@ if [ $stage -le 1 ]; then
 
   mkdir -p $g2p_dir/log
-  auto_vocab_splits=$(eval "echo $auto_vocab_prefix.{$(seq -s',' $nj)}")
+  auto_vocab_splits=$(for n in $(seq $nj); do echo $auto_vocab_prefix.$n; done)
```
also this... revert anything that doesn't have to do with brace expansion.  the bug we are trying to fix is with brace expansion.

danpovey commented on egs/wsj/s5/steps/nnet2/get_egs_discriminative2.sh:
```diff
@@ -91,16 +86,16 @@ utils/split_data.sh $data $nj
   ali_rspecifier="ark,s,cs:gunzip -c $alidir/ali.JOB.gz |"
-  all_ids=$(seq -s, $nj)
-  prior_ali_rspecifier="ark,s,cs:gunzip -c $alidir/ali.{$all_ids}.gz | copy-int-vector ark:- ark,t:- | utils/filter_scp.pl $dir/priors_uttlist | ali-to-pdf $alidir/final.mdl ark,t:- ark:- |"
+  alis=$(for all_ids in $(seq $nj); do echo $alidir/ali.$all_ids.gz; done)
```
this variable name all_ids is not suitable now that you changed the code.  n would be better.  please make the same change elsewhere if you did the same.

LvHang commented:
@danpovey 
I have already reverted the changes in "librispeech" folder.
I have already fixed the variable name in everywhere.
Please check wheather the changes meet your standards. 
Thank you for your time and your suggestion.

Hang

danpovey commented:
Looks OK but can you rebase so there is only one commit?
Something like
git rebase -i HEAD~5
and change all but the first 'pick' to 's' [for squash].
Save your work in a different branch first, in case of errors.

LvHang commented:
@danpovey 
The changes have been integrated. There is only one commit.
Please check it. 
Thank you for your time and your guidance.

Hang
